src/fdc3/js/composeui-fdc3/src/infrastructure/ComposeUIPrivateChannel.ts

```diff
@@ -0,0 +1,197 @@
+
+        return super.broadcast(context)
+    }
+
+    public addContextListener(contextType: string | null, handler: ContextHandler): Promise<Listener>;
+    public addContextListener(handler: ContextHandler): Promise<Listener>;
+    public async addContextListener(contextType: any, handler?: any): Promise<Listener> {
+        var listener = <ComposeUIContextListener>(await super.addContextListener(contextType, handler));
+        listener.setUnsubscribeCallback(l => this.removeContextHandler(l));
+        this.contextHandlers.push(listener);
+
+        this.fireContextHandlerAdded(contextType ?? undefined);
+
+        return listener;
+    }
+
+    private internalEventsHandler(message: TopicMessage) {
+        if (
+            this.disconnected
+            || message.context.sourceId == this.messageRouterClient.clientId
+            || !message.payload) {
+            return;
+        }
+
+        const event = <Fdc3PrivateChannelInternalEvent>JSON.parse(message.payload);
+        switch (event.event) {
+            case "contextListenerAdded":
+                this.addContextListenerHandlers.forEach(handler => {
+                    handler.execute(event.contextType);
+                });
+                break;
+            case "unsubscribed":
+                this.unsubscribeHandlers.forEach(handler =>
+                    handler.execute(event.contextType));
+                break;
+            case "disconnected":
+                this.disconnectHandlers.forEach(handler =>
+                    handler.execute());
+                break;
+        }
+    }
+
+    private fireContextHandlerAdded(contextType: string | undefined) {
+        this.messageRouterClient.publish(this.internalEventsTopic, JSON.stringify(new Fdc3PrivateChannelInternalEvent("contextListenerAdded", contextType)));
+    }
+
+    private removeAddContextListenerHandler(listener: PrivateChannelContextListenerEventListener) {
+        var idx = this.addContextListenerHandlers.indexOf(listener);
+        this.addContextListenerHandlers.splice(idx)
+    }
+
+    private removeUnsubscribeListenerHandler(listener: PrivateChannelContextListenerEventListener) {
+        var idx = this.unsubscribeHandlers.indexOf(listener);
+        this.unsubscribeHandlers.splice(idx)
+    }
+
+    private removeDisconnectListenerHandler(listener: PrivateChannelDisconnectEventListener) {
+        var idx = this.disconnectHandlers.indexOf(listener);
+        this.disconnectHandlers.splice(idx);
+    }
+
+    private async executeForRemoteContextHandlers(handler: PrivateChannelContextListenerEventListener): Promise<void> {
+        let remoteListeners: Array<string | undefined>;
+        try {
+            var listeners = await this.messageRouterClient.invoke(this.remoteContextListenersService, "{}");
+
+            remoteListeners = listeners ? JSON.parse(listeners) : [];
+        }
+        catch (e) {
+            // If this fails, the other side didn't connect yet, so nothing to do
```

Comment:
should we log something here though?

Reply:
This is a normal occurance when setting up the channel for the first time and it has no impact on the local behavior so I wouldn't log it. I'm open for further discussion on this of course :)